### PR TITLE
GitHub App creation flow in `iambic wizard`

### PR DESCRIPTION
**Pull request by smoy**
## What changed?
* `iambic wizard` has a newly added option to setup AWS Lambda GitHub Integration 

## Rationale
* We want a low friction way to setup AWS Lambda GitHub Integration.

## How was it tested?
If it was manually verified, list the instructions for your reviewers to follow.
- [ ] Unit Tests
- [ ] Functional Tests
- [x] Manually Verified

1. The user must be a GitHub org admin; otherwise, they won't be able to create The GitHub App in their GitHub organization.
2. The user must be AWS admin access to deploy Lambda Integration into the target AWS account.

## Architecture

Slide1

### CODE REVIEW

File: deployment/github_app/vars.tf [hunk 1]
```diff
@@ -5,6 +5,60 @@ variable "aws_region" {
...
+  default = ""
+}
+
+variable "name_suffix" {
+  type = string
+  description = "useful for testing to append suffix to resource names"
+  default = ""
+}
+
+variable "use_api_gateway_insetad_of_lambda_functions_url" {
```
Comment (smoy) [inline deployment/github_app/vars.tf:20]: this is where we can switch between API gateway vs lambda function url

**Comment by castrapel** [comment]
Can we try to open the user's browser and print out the URL they should browse to in the terminal?

image

Here's how we do it in our noq-cli - presuming we just need equivalent in python:

```
func popBrowserWindowWithUrl(url string) {
	var openUrlCommand []string = nil
	switch runtime.GOOS {
	case "darwin":
		openUrlCommand = []string{"open"}
	case "linux":
		if isWSL() {
			openUrlCommand = []string{"cmd.exe", "/C", "start"}
		} else {
			openUrlCommand = []string{"xdg-open"}
		}
	case "windows":
		openUrlCommand = []string{"cmd", "/C", "start"}
	}

	if openUrlCommand != nil {
		cmd := exec.Command(openUrlCommand[0], append(openUrlCommand[1:], url)...)
		err := cmd.Start()
		if err == nil {
			err = cmd.Wait()
		}
		if err != nil {
			logging.Log.Errorf("Failed to open browser with '%s': %s.",
				openUrlCommand[0], err.Error())
			logging.Log.Infoln("*** Could not launch browser window.  Open the above link manually to continue. ***")
		} else {
			logging.Log.Infoln(
				"Validation opened in a new browser window. ",
				"Please check your browser for further authentication steps.",
			)
			logging.Log.Infoln(
				"Alternatively, use this URL to authenticate manually: ",
				url,
			)
		}
	} else {
		logging.Log.Infoln("Please open the above URL in a browser and authenticate.")
	}
}
```

**Comment by castrapel** [comment]
Needs spacing 
image

**Comment by castrapel** [comment]
Exception:

```
? Target account id 940552945933
2023/07/31 13:41:14 [info     ] Using AWS profile from environment 
  profile=staging/staging_admin
? Please specify the profile to use to access to the AWS Account.
If None is selected the AWS Account will be skipped. prod/prod_admin

(Optional) Provide a role arn that CloudFormation will assume to create the stack(s) or hit enter to use your current access.
? (Optional) Role ARN
2023/07/31 13:43:34 [info     ] Waiting for stack to be created. 
  estimated_completion=07/31/2023, 13:46:34 
  stack_name=IAMbicGitHubAppRoles-STEVEN-FIXME
2023/07/31 13:44:05 [info     ] Waiting for stack to be created. 
  estimated_completion=07/31/2023, 13:47:05 
  stack_name=IAMbicGitHubAppECRPullThroughCache-STEVEN-FIXME
2023/07/31 13:44:36 [info     ] Waiting for stack to be created. 
  estimated_completion=07/31/2023, 13:47:36 
  stack_name=IAMbicGitHubAppECRRepo-STEVEN-FIXME
2023/07/31 13:45:08 [info     ] Waiting for stack to be created. 
  estimated_completion=07/31/2023, 13:48:08 
  stack_name=IAMbicGitHubAppCodeBuild-STEVEN-FIXME
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 317, in run_module
    run_module_as_main(options.target, alter_argv=True)
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 238, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/Users/ccastrapel/localrepos/iambic/iambic/main.py", line 548, in <module>
    cli()
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/ccastrapel/localrepos/iambic/iambic/main.py", line 532, in setup
    ConfigurationWizard(repo_dir, is_more_options=is_more_options).run()
  File "/Users/ccastrapel/localrepos/iambic/iambic/config/wizard.py", line 2175, in run
    self.configuration_github_app_aws_lambda_setup()
  File "/Users/ccastrapel/localrepos/iambic/iambic/config/wizard.py", line 1989, in configuration_github_app_aws_lambda_setup
    response = code_build_client.start_build(
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/botocore/client.py", line 530, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/botocore/client.py", line 964, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.ResourceNotFoundException: An error occurred (ResourceNotFoundException) when calling the StartBuild operation: Project cannot be found: arn:aws:codebuild:us-east-1:940552945933:project/iambic_code_build
```

**Comment by castrapel** [comment]
Ran into a Codebuild exception with these prompts:

```
✗  /usr/bin/env /Users/ccastrapel/localrepos/iambic/venv/bin/python /Users/ccastrapel/.vscode/extensions/ms-python.pyth
on-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher 63060 -- -m iambic.main setup 
2023/08/02 05:08:06 [info     ] Found existing configuration file 
  config_path=/Users/ccastrapel/localrepos/noq-templates/config/noq_config.yaml
2023/08/02 05:08:06 [info     ] Loading config...
2023/08/02 05:08:06 [info     ] Setting config metadata...
2023/08/02 05:08:12 [info     ] Plugins loaded successfully...
? What would you like to configure? Setup GitHub App Integration using AWS Lambda

IAMbic detected you are using arn:aws:iam::259868150464:role/staging_admin for AWS access.
This identity will require the ability to create CloudFormation stacks, stack sets, and stack set instances.
? Would you like to use this identity? Yes
2023/08/02 05:08:35 [info     ] Scanning for upstream changes to config attributes.
2023/08/02 05:08:37 [info     ] Finished scanning for upstream changes to config attributes.
2023/08/02 05:08:37 [info     ] Config successfully written    
  config_location=/Users/ccastrapel/localrepos/noq-templates/config/noq_config.yaml

This requires that you have the ability to create CloudFormation stacks

? Proceed? Yes

To setup GitHub App for iambic it requires creating CloudFormation stacks. 
To review the templates used or deploy them manually, the IdentityRule templates used can be found here:
https://github.com/noqdev/iambic/tree/main/iambic/plugins/v0_1_0/aws/cloud_formation/templates
If you have already manually deployed the templates, answer yes to proceed.
IAMbic will validate that your stacks have been deployed successfully and will not attempt to replace them.
? Add Tags (leave blank or `team=ops_team, cost_center=engineering`):
? Proceed? Yes
? We recommend you deploy Lambda integration on a non-management account.
Target AWS Account name:  demo-3
2023/08/02 05:13:41 [info     ] Target AWS Account ID is 518317429440
2023/08/02 05:13:41 [info     ] Using AWS profile from environment 
  profile=staging/staging_admin
? Please specify the profile to use to access to the AWS Account.
If None is selected the AWS Account will be skipped. demo-3/demo3_admin
2023/08/02 05:13:49 [info     ] Waiting for stack to be created. 
  estimated_completion=08/02/2023, 05:16:49 
  stack_name=IAMbicGitHubAppRoles
2023/08/02 05:14:20 [info     ] Added trust policy on arn:aws:iam::259868150464:role/IambicHubRole for arn:aws:iam::518317429440:role/iambic_github_app_lambda_execution
2023/08/02 05:14:21 [info     ] Waiting for stack to be created. 
  estimated_completion=08/02/2023, 05:17:21 
  stack_name=IAMbicGitHubAppECRPullThroughCache
2023/08/02 05:14:52 [info     ] Waiting for stack to be created. 
  estimated_completion=08/02/2023, 05:17:52 
  stack_name=IAMbicGitHubAppECRRepo
2023/08/02 05:15:23 [info     ] Waiting for stack to be created. 
  estimated_completion=08/02/2023, 05:18:23 
  stack_name=IAMbicGitHubAppCodeBuild
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 317, in run_module
    run_module_as_main(options.target, alter_argv=True)
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 238, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/ccastrapel/.vscode/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/Users/ccastrapel/localrepos/iambic/iambic/main.py", line 541, in <module>
    cli()
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/ccastrapel/localrepos/iambic/iambic/main.py", line 525, in setup
    ConfigurationWizard(repo_dir, is_more_options=is_more_options).run()
  File "/Users/ccastrapel/localrepos/iambic/iambic/config/wizard.py", line 2286, in run
    self.configuration_github_app_aws_lambda_setup()
  File "/Users/ccastrapel/localrepos/iambic/iambic/config/wizard.py", line 2041, in configuration_github_app_aws_lambda_setup
    self.github_app_pull_latest_iambic_image(session)
  File "/Users/ccastrapel/localrepos/iambic/iambic/config/wizard.py", line 2123, in github_app_pull_latest_iambic_image
    response = code_build_client.start_build(
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/botocore/client.py", line 530, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/ccastrapel/localrepos/iambic/venv/lib/python3.10/site-packages/botocore/client.py", line 964, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.ResourceNotFoundException: An error occurred (ResourceNotFoundException) when calling the StartBuild operation: Project cannot be found: arn:aws:codebuild:us-east-1:518317429440:project/iambic_code_build
```

File: iambic/config/wizard.py [hunk 2]
```diff
@@ -1104,7 +1111,7 @@ def setup_aws_configuration(self):
             identity_arn = get_identity_arn(default_caller_identity)
             click.echo(
                 f"\nIAMbic detected you are using {identity_arn} for AWS access.\n"
-                f"This identity will require the ability to create"
+                f"This identity will require the ability to create "
```
Comment (castrapel) [inline iambic/config/wizard.py:1114]: Should it also have the ability to create AWS secrets?

File: iambic/plugins/v0_1_0/github/create_github_app.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def run_local_webserver():
+    app = web.Application()
+    app.add_routes(routes)
+    web.run_app(app, host="localhost", port=8080, print=False)
+
+
+def get_github_app_secrets():
+    full_path = os.path.expanduser("~/.iambic")
```
Comment (castrapel) [inline iambic/plugins/v0_1_0/github/create_github_app.py:61]: SAVE_DIR

**Pull request merged**